src/Processors/QueryPlan/Optimizations/distinctReadInOrder.cpp

```diff
@@ -36,6 +105,11 @@ size_t tryDistinctReadInOrder(QueryPlan::Node * parent_node)
 
         steps_to_update.push_back(step);
 
+        if (const auto * const expr = typeid_cast<const ExpressionStep *>(step); expr)
+            dag_stack.push_back(expr->getExpression());
+        else if (const auto * const filter = typeid_cast<const FilterStep *>(step); filter)
+            dag_stack.push_back(filter->getExpression());
```

Comment:
I think we need to add a kind of allow list of steps here as well.
In general, a step can change columns as it likes, and `mergeInplace` can throw an exception in case if inputs of one DAG are not found in the outputs of another.